### PR TITLE
feat(ecmascript): Implement the `Function` constructor

### PR DESCRIPTION
**Pull request by andreubotella**
This patch implements the `Function`, `AsyncFunction` and `GeneratorFunction` constructors, which parse the arguments as source code and build a function from it.

It seems like this addition marks the first time that a function can be constructed whose source text is not part of the ECMAScript code context's source code. Therefore, we add an optional `SourceCode` argument to `OrdinaryFunctionCreateParams` so the stringification works properly.

One issue in the implementation of these functions is that the oxc parser's API only exposes entrypoints to parse a whole script/module, or an expression; while the spec for these constructors requires that the function arguments and the function body be parsed separately, to avoid injection.

Due to oxc not exposing this API, we can't follow the spec exactly here. What we do instead is generate the spec-mandated source code for the function (that would be returned when stringified), parse that as a script, and throw if the AST doesn't match exactly one statement which is a function declaration of the right type.

### CODE REVIEW

File: nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs [hunk 1]
```diff
@@ -52,3 +74,172 @@ impl FunctionConstructor {
...
+                        });
+                    }
+                }
+            }
+        }
+
+        if let Some(function) = function {
+            (function, source_code.unwrap())
+        } else {
+            return Err(agent.throw_exception_with_static_message(
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs:180]: **nitpick:** We haven't really been doing this but we should and could: We know that SourceCode cannot be referred to by any other anywhere, and that it is the last item in the source_codes Vec. We can thus pop it off the Vec before returning, so as to help GC along.

File: nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs [hunk 2]
```diff
@@ -52,3 +74,172 @@ impl FunctionConstructor {
...
+    };
+    let f = ordinary_function_create(agent, params);
+
+    set_function_name(
+        agent,
+        f,
+        BUILTIN_STRING_MEMORY.anonymous.to_property_key(),
+        None,
+    );
+    match kind {
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs:212]: **suggestion:** Move these branches to the callers, as they statically know what their kind is.

**Review by aapoalas — APPROVED** [review]
Niiiice <3

File: nova_vm/src/ecmascript/builtins/ecmascript_function.rs [hunk 3]
```diff
@@ -815,10 +816,17 @@ pub(crate) fn ordinary_function_create<'agent, 'program>(
...
+    } else {
+        let running_ecmascript_code = &agent.running_execution_context().ecmascript_code.unwrap();
+        (
+            running_ecmascript_code.source_code,
+            running_ecmascript_code.is_strict_mode,
+        )
+    };
     // 7. If the source text matched by Body is strict mode code, let Strict be true; else let Strict be false.
-    let strict = params.body.has_use_strict_directive() || running_ecmascript_code.is_strict_mode;
+    let strict = params.body.has_use_strict_directive() || outer_env_is_strict;
```
[the remark this answers is not in the document]
Comment (andreubotella) [reply nova_vm/src/ecmascript/builtins/ecmascript_function.rs]: Done.

File: nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs [hunk 4]
```diff
@@ -52,3 +74,172 @@ impl FunctionConstructor {
...
+    parameter_args: &[Value],
+    body_arg: Value,
+) -> JsResult<ECMAScriptFunction> {
+    // 11. Perform ? HostEnsureCanCompileStrings(currentRealm, parameterStrings, bodyString, false).
+    agent
+        .host_hooks
+        .host_ensure_can_compile_strings(agent.current_realm_mut())?;
+
+    let parameters = {
+        let mut parameters = std::string::String::new();
```
[the remark this answers is not in the document]
Comment (andreubotella) [reply nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs]: Done.

File: nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs (hunk 1 above)
Comment (andreubotella) [reply nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs:212]: Done.

File: nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs (hunk 2 above)
Comment (andreubotella) [reply nova_vm/src/ecmascript/builtins/fundamental_objects/function_objects/function_constructor.rs]: Done.

File: nova_vm/src/ecmascript/builtins/control_abstraction_objects/generator_function_objects/generator_function_constructor.rs [hunk 5]
```diff
@@ -46,15 +53,46 @@ impl GeneratorFunctionConstructor {
...
+            Some(
+                agent
+                    .current_realm()
+                    .intrinsics()
+                    .generator_prototype()
+                    .into_object(),
+            ),
+        );
+        //   b. Perform ! DefinePropertyOrThrow(F, "prototype", PropertyDescriptor { [[Value]]: prototype, [[Writable]]: true, [[Enumerable]]: false, [[Configurable]]: false }).
+        define_property_or_throw(
```
[the remark this answers is not in the document]
Comment (andreubotella) [reply nova_vm/src/ecmascript/builtins/control_abstraction_objects/generator_function_objects/generator_function_constructor.rs:80]: This is setting the `prototype` property on the function to return, not on the newly-created prototype.

[the remark this answers is not in the document]
Comment (aapoalas) [reply nova_vm/src/ecmascript/builtins/control_abstraction_objects/generator_function_objects/generator_function_constructor.rs:80]: Oh, oops, yeah, sorry <3

**Pull request merged**